### PR TITLE
Fix Assertion Error Message

### LINKED ISSUE #1333: Failing Assertion is printing "assertion not found" not presented.

### PR DESCRIPTION
**Pull request by willsaid-google**
Fixes #1333

### CODE REVIEW

File: AppFramework/Assertion/GREYAssertions.m [hunk 1]
```diff
@@ -59,7 +62,7 @@ @implementation GREYAssertions
     NSString *matcherDiagnosticsID = [matcher diagnosticsID];
     if (matcherDiagnosticsID) {
       NSString *assertionDiagnosticsID =
-          [NSString stringWithFormat:@"assertWithMatcher:%@", matcherDiagnosticsID];
+      [NSString stringWithFormat:@"assertWithMatcher:%@", matcherDiagnosticsID];
```
Comment (tirodkar) [inline AppFramework/Assertion/GREYAssertions.m:65]: Fix formatting.

File: AppFramework/Assertion/GREYAssertions.m [hunk 2]
```diff
@@ -39,14 +39,17 @@ @implementation GREYAssertions
   GREYCheckBlockWithError assertionBlock = ^BOOL(id element, __strong NSError **errorOrNil) {
     GREYStringDescription *mismatch = [[GREYStringDescription alloc] init];
     if (![matcher matches:element describingMismatchTo:mismatch]) {
-      NSMutableString *reason = [[NSMutableString alloc] init];
       if (!element) {
-        [reason appendFormat:@"Assertion with Matcher failed: No UI element was matched.\n"];
-        [reason appendFormat:@"Matcher: \n%@", [matcher description]];
+        NSString *reason = [NSString stringWithFormat:
```
Comment (tirodkar) [inline AppFramework/Assertion/GREYAssertions.m:43]:
```
Exception Name: com.google.earlgrey.ElementInteractionErrorDomain
Exception Reason: 
Matched UI Element: <UITableViewLabel:0x7ff259d28540; isAccessible=Y; AX.label='Basic Views'; AX.frame={{20, 308}, {374, 44}}; AX.activationPoint={207, 330}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{20, 0}, {374, 44}}; opaque; alpha=1; UIE=N; text='Basic Views'>
Did not match: isNil
In assertion matcher: isNil
Exception Details: An assertion failed.
Exception with Assertion: {
  "Assertion Criteria" : "assertWithMatcher:isNil",
  "Element Matcher" : "((kindOfClass('UILabel') || kindOfClass('UITextField') || kindOfClass('UITextView')) && hasText('Basic Views'))"
}
```

Do you think we can format the `Exception Reason:` better? I don't like how it's all in one text blob.

We do not have test cases for the assertion description / info so can we add them ErrorHandlingTest?

Comment (willsaid-google) [reply AppFramework/Assertion/GREYAssertions.m:43]: New output:
```Exception Name: com.google.earlgrey.ElementInteractionErrorDomain
Exception Reason: The following elements do not match the element matcher: isNil 
  1. <UIWindow:0x7f7fac711270; isAccessible=N; AX.frame={{0, 0}, {375, 812}}; AX.activationPoint={187.5, 406}; AX.traits='UIAccessibilityTraitNone'; AX.focused='N'; frame={{0, 0}, {375, 812}}; opaque; alpha=1> 
  2. isNil 

Exception Details: An assertion failed.
Exception with Assertion: {
  "Assertion Criteria" : "assertWithMatcher:isNil",
  "Element Matcher" : "(kindOfClass('UIWindow') && keyWindow)"
}```

Comment (tirodkar) [reply AppFramework/Assertion/GREYAssertions.m:43]: Can we have 

Element: <UIWindow..
Did not match: isNil ?

Comment (willsaid-google) [reply AppFramework/Assertion/GREYAssertions.m:43]: Then what about the Element Matcher?

Comment (willsaid-google) [reply AppFramework/Assertion/GREYAssertions.m:43]: Resolved: 
```
Exception Reason: Element does not fulfill assertion criteria: isNil 
Element: <UIWindow:0x7fc1f6c13940; isAccessible=N; AX.frame={{0, 0}, {375, 812}}; AX.activationPoint={187.5, 406}; AX.traits='UIAccessibilityTraitNone'; AX.focused='N'; frame={{0, 0}, {375, 812}}; opaque; alpha=1> 
Mismatch: isNil 

Exception Details: An assertion failed.
Exception with Assertion: {
  "Assertion Criteria" : "assertWithMatcher:isNil",
  "Element Matcher" : "(kindOfClass('UIWindow') && keyWindow)"
}
```

File: Tests/Functional/Sources/ErrorHandlingTest.m [hunk 3]
```diff
@@ -236,6 +236,15 @@ - (void)testKeyboardDismissalError {
   XCTAssertTrue([error.description containsString:keyboardErrorString]);
 }
 
+/// kGREYInteractionAssertionFailedErrorCode
```
Comment (tirodkar) [inline Tests/Functional/Sources/ErrorHandlingTest.m:239]: We try to comment new / modified tests as we proceed.

/**
 * Verifies the printed information for a GREYError found for an assertion failure.
 */

File: Tests/Functional/Sources/ErrorHandlingTest.m [hunk 4]
```diff
@@ -236,6 +236,15 @@ - (void)testKeyboardDismissalError {
...
+- (void)testAssertionInteractionError {
```
Comment (tirodkar) [inline Tests/Functional/Sources/ErrorHandlingTest.m:240]: testAssertionFailureDescription

File: AppFramework/Assertion/GREYAssertions.m [hunk 5]
```diff
@@ -39,14 +39,17 @@ @implementation GREYAssertions
...
+                            @"Assertion with Matcher failed: No UI element was matched.\n"
+                            @"Matcher: \n%@", [matcher description]];
         I_GREYPopulateError(errorOrNil, kGREYInteractionErrorDomain,
                             kGREYInteractionElementNotFoundErrorCode, reason);
       } else {
-        [reason appendFormat:@"An assertion failed."];
+        NSString *reason = [NSString stringWithFormat:@"Element does not fulfill "
```
Comment (tirodkar) [inline AppFramework/Assertion/GREYAssertions.m:49]: Format as such:
```
NSString *reason =
    [NSString stringWithFormat:@"Element does not fulfill assertion criteria:%@ "
                               @"\nElement: %@ \nMismatch: %@ \n",
                               [matcher description], [element grey_description], [mismatch description]];
```

Formatting on github seems off, ensure the 100ch line length is filled per line.

**Comment by tirodkar** [comment]
Closing in favor of https://github.com/google/EarlGrey/pull/1336.

**Pull request closed without merging**